src/node.cc

```diff
@@ -3394,13 +3403,23 @@ int EmitExit(Environment* env) {
   HandleScope handle_scope(env->isolate());
   Context::Scope context_scope(env->context());
   Local<Object> process_object = env->process_object();
-  process_object->Set(FIXED_ONE_BYTE_STRING(env->isolate(), "_exiting"),
-                      True(env->isolate()));
+
+  Local<String> _exiting_string =
+      FIXED_ONE_BYTE_STRING(env->isolate(), "_exiting");
+  Local<Value> exiting = process_object->Get(_exiting_string);
+
+  if (exiting->IsFalse()) {
+    process_object->Set(_exiting_string, True(env->isolate()));
+  }
 
   Local<String> exit_code = env->exit_code_string();
   int code = process_object->Get(env->context(), exit_code).ToLocalChecked()
       ->Int32Value(env->context()).ToChecked();
 
+  if (exiting->IsTrue()) {
+    return code;
+  }
```

Comment:
I believe you when you say the test was failing before, but do you know why this was happening? Does it point to a larger issue?

Reply:
As I understand the issue here is that there is no synchronization between worker-exit and worker-fatalException-exit (links at the end of the first post) and as a result we have 
* fatal callback called in js land -> emit exit from there
* worker undestands that it is finishing, exits its loop and calls `EmitExit`
Therefore we get 2 calls to 'exit' listeners. I'm not sure if 'just muting' second exit event is a good solution, hence my note in the end.

> I believe you when you say the test was failing before

I'm not sure what you mean? Also I forgot to add a description of a test case for the bug (it is [here](https://github.com/nodejs/node/pull/21739/files#diff-336ff635d2ed9bcd3a676887181dd1baR19) in the tests and I'll update the first post soon)

Reply:
to be honest ,you are really so konwledgable

Reply:
@lundibundi I see, that makes sense. I think I’d have a slight preference to handle this using a flag on the `Environment` object, since `process._exiting` can just be modified by userland code.

Reply:
Oh, that's surely better. I'll try to implement that, though there may be some problems as `_exiting` is actually set from `_fatalException` handler in js, so with `_exiting` it is set `if no handler and **before** 'exit' callback`, but with env it will be set `if no handler and **after** all 'exit' callbacks` due to the way fatalException is handled in js.

Reply:
@lundibundi I think there’s a bigger bug here – in `test-worker-uncaught-exception.js`, the worker does not exit because of the uncaught exception, but because there’s no work to do after it. :/

I think we want a `process.exit()` call at the end of the `if (!caught) {` block?

Reply:
@addaleax oh, that is indeed true, great catch thanks.
I think adding a Timeout? with assert.fail should be good enough to ensure that the worker exits. Though I'm not sure if timeout is a good idea (flakiness and such), is there any better way?
Yeah `process.exit()` seems to work just fine there.

Reply:
@lundibundi We could try to start some async operation before the uncaught `throw` and see whether it’s executed? That latter part might be tricky, but that would be the general idea, I think